docs: charmcraft fetch-libs deprecation (#491)
canonical/charmlibs: .docs/explanation/charmhub-libraries-deprecation.md

> (charmhub-libraries-deprecation)=

Reviewer: Don't forget to add a meta description at the beginning of the file

Author: Oops, thanks!

# Charmhub-hosted libraries are being phased out

Charmhub-hosted charm libraries -- single-file Python modules fetched with `charmcraft fetch-libs` -- are being phased out in favour of standard Python packages distributed on PyPI.

## Background

Charms originally shared code by distributing single-file Python modules on Charmhub. Each module was associated with a specific charm. Charms fetched libraries using the `charmcraft fetch-lib` and `charmcraft fetch-libs` commands, and typically checked them into their version control. These libraries were stored under `lib/`, which was added to `PYTHONPATH` at runtime, allowing the module to be imported by charm name and API version. For example `from operator_libs_linux.v2 import snap`.

## Timeline

- `charmcraft` emits a deprecation warning on library operations (we are here).

> - Charmhub disables uploading *new* libraries (26.04 cycle).

Reviewer: Is 26.04 correct? The cycle having just ended

Author: Good catch!
```suggestion
- Charmhub disables uploading *new* libraries (26.10 cycle).
```

- Charmhub disables updates to existing libraries.

## Why the change?

### No dependency resolution

> Charmhub-hosted libraries have no mechanism for declaring or resolving dependencies on other Python packages. If a library needs a dependency, it lists the package name in a `PYDEPS` variable, and the charm author must manually add each dependency to their charm's own requirements. That is, transitive dependencies of Charmhub-hosted libraries are tracked manually, and the charm author is also responsible for resolving a compatible version for the different libraries that might require that dependency.

Reviewer: `PYDEPS` feels like declaring, even if it's for humans. I think this would be fine:

```suggestion
Charmhub-hosted libraries have no mechanism for resolving dependencies on other Python packages. If a library needs a dependency, it lists the package name in a `PYDEPS` variable, and the charm author must manually add each dependency to their charm's own requirements. That is, transitive dependencies of Charmhub-hosted libraries are tracked manually, and the charm author is also responsible for resolving a compatible version for the different libraries that might require that dependency.
```

Standard Python packages, by contrast, declare their dependencies in metadata that tools like `pip` and `uv` resolve automatically.

### Vendored code is hard to maintain

Because Charmhub libraries are committed into each charm's repository, updating a library means re-running `charmcraft fetch-libs` and committing the result. Across many charms, this creates a maintenance burden and increases the risk of running outdated code.

### Limited tooling integration

Charmhub libraries do not participate in standard Python tooling. This makes it trickier to ensure their visibility for dependency scanners and vulnerability checkers, leads to the library itself not being visible in the charm's lock file, and requires extra work to make the library visible in IDEs.

## What replaces it?

Charm libraries should be distributed as regular Python packages, typically on PyPI.
Libraries of broad interest are published under the `charmlibs` namespace from the [`charmlibs` monorepo](https://github.com/canonical/charmlibs). Team-specific libraries can be distributed on PyPI under their own namespace, as Git dependencies, or as local packages in a charm monorepo.

> > See {ref}`how-to-python-package` for full details on distribution options.

Reviewer: I don't think we need the surrounding formatting and text. (Also applies to the links at the end of other sections)

```suggestion
See {ref}`how-to-python-package`
```

Author: Ah yes, I forgot we planned to drop block quotes for things like this. I do wish we had a standard "see more", "read next" kind of component though. Is "see X" with no other punctuation the way?

Author: Maybe "Read more: X" ?

Author: I went with "read more:"

## What should I do?

> **If you maintain a Charmhub-hosted library:**

Reviewer: I'd drop the colons. Or switch to subheadings

Author: ```suggestion
### If you maintain a Charmhub-hosted library
```

> Consider migrating it to a Python package. If it implements a widely-used interface or provides broadly useful functionality, it belongs in the `charmlibs` monorepo.

Reviewer: If we're ending support, we definitely want people to migrate, don't we? Even if it's not distributed on PyPI, it should still be a Python package.

```suggestion
Migrate it to a Python package. If it implements a widely-used interface or provides broadly useful functionality, it belongs in the `charmlibs` monorepo.
```

> See {ref}`how-to-migrate` for a step-by-step guide.

> **If you use Charmhub-hosted libraries in your charm:**

Author: ```suggestion
### If you use Charmhub-hosted libraries in your charm
```

When a Python package replacement is available, switch to it:

> 1. Remove the library from the `charm-libs` section of your `charmcraft.yaml` and delete the vendored file from `lib/`.

Reviewer: Directory?

```suggestion
1. Remove the library from the `charm-libs` section of your `charmcraft.yaml` and delete the vendored directory from `lib/`.
```

2. Remove any manually added transitive dependencies that are no longer needed.

> 3. Add the replacement package to your charm's dependencies (e.g. in `pyproject.toml`).

Reviewer: I think we can drop the last part, as the linked doc covers it in more detail.

```suggestion
3. Add the replacement package to your charm's dependencies.
```

4. Update your imports to use the new package's import path.

> See {ref}`how-to-manage-charm-libraries` for guidance on managing Python package dependencies in your charm.

> **If you are writing a new library:**

Author: ```suggestion
### If you are writing a new library
```
